core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushProjectionThroughJoinIntoTableScan.java

```diff
@@ -0,0 +1,246 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner.iterative.rule;
+
+import com.google.common.collect.ImmutableSet;
+import io.trino.matching.Capture;
+import io.trino.matching.Captures;
+import io.trino.matching.Pattern;
+import io.trino.sql.ir.Expression;
+import io.trino.sql.ir.FieldReference;
+import io.trino.sql.ir.Reference;
+import io.trino.sql.planner.DeterminismEvaluator;
+import io.trino.sql.planner.PlanNodeIdAllocator;
+import io.trino.sql.planner.Symbol;
+import io.trino.sql.planner.SymbolsExtractor;
+import io.trino.sql.planner.iterative.Rule;
+import io.trino.sql.planner.plan.Assignments;
+import io.trino.sql.planner.plan.JoinNode;
+import io.trino.sql.planner.plan.JoinType;
+import io.trino.sql.planner.plan.PlanNode;
+import io.trino.sql.planner.plan.ProjectNode;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Stream;
+
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static io.trino.matching.Capture.newCapture;
+import static io.trino.sql.planner.SymbolsExtractor.extractUnique;
+import static io.trino.sql.planner.plan.Patterns.join;
+import static io.trino.sql.planner.plan.Patterns.project;
+import static io.trino.sql.planner.plan.Patterns.source;
+
+/**
+ * Pushes projections that appear above a join down to the table scans on either side of the join.
+ * This is particularly useful for cross-connector joins where the join cannot be pushed down, but
+ * individual projections on each side can still be pushed to their respective connectors.
+ *
+ * <pre>
+ * Transforms:
+ *   Project(x := f(a), y := g(b))
+ *     Join(a = b)
+ *       TableScan(a, ...)
+ *       TableScan(b, ...)
+ *
+ * Into:
+ *   Project(x, y)  -- identity projections
+ *     Join(a = b)
+ *       Project(x := f(a), a)  -- pushed down
+ *         TableScan(a, ...)
+ *       Project(y := g(b), b)  -- pushed down
+ *         TableScan(b, ...)
```

Comment:
But if the `Join` filters the data then these projection would be redundant right ?